### PR TITLE
[FLINK-29337][hive] Fix fail to query non-hive table in Hive dialect 

### PR DESCRIPTION
**Pull request by luoyuxia**
## What is the purpose of the change
Ti fix the failure for the query with non-hive table in Hive dialect.

## Brief change log
- Modify [FromClauseASTParser.g](https://github.com/apache/flink/pull/20855/files#diff-2c1c44e0dd96a52613a64a23d27f68596207577457d5a0d2cbc9f1447daeba0f) to make table can be `catalog.db.table`.
- Parse table name user specified to `catalog.db.table` intead of `db.table`.
- Replace the Hive's `Table` which only represents `db.table` with Flink's `CatalogTable` which can  represents `catalog.db.table`.

## Verifying this change
Added test in [HiveDialectQueryITCase.java](https://github.com/apache/flink/pull/20855/files#diff-fb5fc4106e27e27f9968e59b643003444fd40af58aa2f147ac301009b53a47ca).

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

**Comment by luoyuxia** [comment]
@flinkbot run azure

**Comment by luoyuxia** [comment]
@flinkbot run azure

**Comment by luoyuxia** [comment]
test for hive3 is in https://github.com/apache/flink/pull/20907

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/copy/HiveParserSemanticAnalyzer.java [hunk 1]
```diff
@@ -2158,6 +2074,44 @@ public String getAutogenColAliasPrfxLbl() {
         return this.autogenColAliasPrfxLbl;
     }
 
+    public CatalogBaseTable getCatalogBaseTable(String tableName, HiveParserQB qb) {
+        return getCatalogBaseTable(tableName, qb, true);
+    }
+
+    public CatalogBaseTable getCatalogBaseTable(
```
Comment (lsyldliu) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/copy/HiveParserSemanticAnalyzer.java:2081]: Nit: add `@Nullable` annotation.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/HiveParserUtils.java [hunk 2]
```diff
@@ -1689,4 +1694,61 @@ public static boolean isFromTimeStampToDecimal(RelDataType srcType, RelDataType
...
+                                        c.getName(),
+                                        (DataType) c.getDataType(),
+                                        c.getMetadataKey(),
+                                        c.isVirtual());
+                            } else if (column instanceof Schema.UnresolvedComputedColumn) {
+                                final Schema.UnresolvedComputedColumn c =
+                                        (Schema.UnresolvedComputedColumn) column;
+                                return TableColumn.computed(
+                                        c.getName(),
+                                        unresolvedType,
```
Comment (lsyldliu) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/HiveParserUtils.java:1728]: Why not use `DataTypes.TIMESTAMP(3)` directly?

**Review by lsyldliu — COMMENTED** [review]
Thanks for contribution, LGTM overall, I left some minor comments.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/HiveParserUtils.java [hunk 3]
```diff
@@ -1689,4 +1694,61 @@ public static boolean isFromTimeStampToDecimal(RelDataType srcType, RelDataType
         return srcType.getSqlTypeName().equals(SqlTypeName.TIMESTAMP)
                 && targetType.getSqlTypeName().equals(SqlTypeName.DECIMAL);
     }
+
+    /**
+     * Helps to migrate the new {@link Schema} to old API methods. HiveCatalog use deprecated {@link
+     * TableSchema}, other catalogs may use the new {@link Schema}. Currently, we use it to unify to
+     * {@link TableSchema}. It should be dropped after dropping {@link TableSchema}.
+     */
+    public static TableSchema fromUnresolvedSchema(Schema schema) {
```
[the remark this answers is not in the document]
Comment (luoyuxia) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/HiveParserUtils.java:1703]: I have created FLINK-29585 for it.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/HiveParserUtils.java (hunk 2 above)
Comment (luoyuxia) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/HiveParserUtils.java:1728]: `DataTypes.TIMESTAMP(3)`  will be used in two places, so I prefer to create a local variable `unresolvedType ` and make these two places use the variable.

**Comment by luoyuxia** [comment]
@flinkbot run azure

**Review by lsyldliu — APPROVED** [review]
LGTM

**Comment by luoyuxia** [comment]
@flinkbot run azure

**Comment by luoyuxia** [comment]
@wuchong Could you please help merge?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/HiveParserUtils.java [hunk 4]
```diff
@@ -1689,4 +1694,61 @@ public static boolean isFromTimeStampToDecimal(RelDataType srcType, RelDataType
...
+        final TableSchema.Builder builder = TableSchema.builder();
+
+        final DataType unresolvedType = DataTypes.TIMESTAMP(3);
```
Comment (Aitozi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/planner/delegation/hive/HiveParserUtils.java:1706]: Hi @luoyuxia , I have a question here, why here unresolvedType is hard coded `TIMESTAMP(3)` ?

**Pull request merged**